python/typemap_utils.cpp

```diff
@@ -474,6 +474,28 @@ static int pymaterial_to_material(PyObject *po, material_type *mt) {
     } else if (IsPyString(po)) {
         const char *eps_input_file = PyObject_ToCharPtr(po);
         md = make_file_material(eps_input_file);
+    } else if (PyArray_Check(po)) {
+        PyArrayObject *pao = (PyArrayObject*)po;
+
+        if (!PyArray_ISCARRAY(pao)) {
+            PyErr_SetString(PyExc_TypeError, "Numpy array must be C-style contiguous.");
+            return 0;
+        }
+        md = (material_data *)malloc(sizeof(*md));
+        md->which_subclass=material_data::MATERIAL_FILE;
+        md->epsilon_dims[0] = md->epsilon_dims[1] = md->epsilon_dims[2] = 1;
+        md->epsilon_data = (double*)PyArray_DATA(pao);
```

Comment:
Won’t this cause problems if the numpy array is garbage-collected before we are done with the material? Or can that never happen here?

Reply:
I think the array should be alive as long as the `Simulation` instance because they have to exist in the same scope. It could be garbage collected in a situation like this:

```py
def init():
    arr = np.array()  # eps data
    sim = mp.Simulation(default_material=arr)
    return sim

sim = init()
sim.run()
sim.reset_meep()
sim.default_material=mp.air  # arr is marked for collection
sim.run()
```
but in that case we're no longer using the `arr` material.

 I was worried that `libmeepgeom` may delete `epsilon_data` when numpy is supposed to own it, but it turns out that never happens. I think this PR is good to go, but I think there is a leak when `epsilon_data` contains the result of `h5file.read` in the `MATERIAL_FILE` case. I'll look into that separately.